TRIB-143: implements correct To Be Reviewed Checker Service

mrsbluerose opened the pull request:
- Deletes unused To Be Reviewed checker class and sets correct service to run every 10 minutes
- Includes refactoring for several encountered bugs
- Fixes tests to run with refactored code

Notable changes:
- Removes all phrase validation. Such checks were redundant with the phrase service providing this validation when phrases are entered and passed to the AttributesAPIController
- Refactors CheckPartOfSpeech method to handle rare case when the Words API has a null result for parts of speech but does have all the other properties (i.e. the word exists and is missing the part of speech definition)
- Refactors ValidatePhrase method to handle when adverb and preposition are defined with "nullvalue"
- Refactors several calls to repositories to use hibernate .save() and .delete() instead of custom queries

mrsbluerose commented on src/main/java/com/savvato/tribeapp/services/ToBeReviewedCheckerService.java:
Refactored the checkPartOfSpeech params to not include call to wordDetails. Reasoning: In trying to handle the oddball case of words API not having part of speech defined for a word, I moved all logic for getting word details in check part of speech.

haxwell commented on src/main/java/com/savvato/tribeapp/services/ToBeReviewedCheckerServiceImpl.java:
```diff
@@ -5,110 +5,127 @@
-            }
-        } catch (Exception e) {
-            LOGGER.warning("The " + expectedPartOfSpeech + " passed in isn't an English word!");
+
```
You don't need this method... In the places where you call this method, you can call checkPartOfSpeech() directly.

mrsbluerose replied:
I'm not sure what method you are referring to.

haxwell replied:
The method validatePhraseComponent(). It just calls checkPartOfSpeech() and returns the return value. You could instead, just call checkPartOfSpeech() wherever you are currently calling validatePhraseComponent().

mrsbluerose replied:
I had to dig in to wrap my brain around this again. I believe the reasoning behind having the validatePhraseComponent() function is to log which word is the problem (or words, if multiple problems). If we just called checkPartOfSpeech() for each && piece of validatePhrase > boolean validPhrase, we get an end result of true or false for the whole phrase. If we want to log any individual words as the problem, we either need to keep validatePhraseComponent() as it is or refactor validatePhrase() to do all the word checks individually instead of in a boolean with multiple &&s. How do you prefer to handle it?

haxwell replied:
Here is a diff describing what I had in mind. (Google 'how to read a diff' for help reading it, if need be)

```jjames@hppav06:~/src/dmpj-folks/mrsbluerose/tribe-app-backend$ gd src/main/java/com/savvato/tribeapp/services/ToBeReviewedCheckerServiceImpl.java
diff --git a/src/main/java/com/savvato/tribeapp/services/ToBeReviewedCheckerServiceImpl.java b/src/main/java/com/savvato/tribeapp/services/ToBeReviewedCheckerServiceImpl.java
index a04c217..467b819 100644
--- a/src/main/java/com/savvato/tribeapp/services/ToBeReviewedCheckerServiceImpl.java
+++ b/src/main/java/com/savvato/tribeapp/services/ToBeReviewedCheckerServiceImpl.java
@@ -86,31 +86,25 @@ public class ToBeReviewedCheckerServiceImpl implements ToBeReviewedCheckerServic
                     return true;
                 }
             }
-            return partsOfSpeech.contains(expectedPartOfSpeech);
-        }
-    }
-
-    @Override
-    public boolean validatePhraseComponent(String word, String expectedPartOfSpeech) {
 
-        Boolean validPartOfSpeech = checkPartOfSpeech(word, expectedPartOfSpeech);
+            boolean contains = partsOfSpeech.contains(expectedPartOfSpeech);
 
-        if (validPartOfSpeech) {
-            return true;
-        } else {
-            log.warn("The word passed in isn't a " + expectedPartOfSpeech + "!");
-            return false;
+            if (contains) {
+                return true;
+            } else {
+                log.warn("The word passed in isn't a " + expectedPartOfSpeech + "!");
+                return false;
+            }
         }
-
     }
 
     @Override
     public void validatePhrase(ToBeReviewed tbr) {
 
-        boolean validPhrase = validatePhraseComponent(tbr.getNoun(), "noun")
-                && validatePhraseComponent(tbr.getVerb(), "verb")
-                && (tbr.getAdverb().equals(Constants.NULL_VALUE_WORD) || validatePhraseComponent(tbr.getAdverb(), "adverb"))
-                && (tbr.getPreposition().equals(Constants.NULL_VALUE_WORD) || validatePhraseComponent(tbr.getPreposition(), "preposition"));
+        boolean validPhrase = checkPartOfSpeech(tbr.getNoun(), "noun")
+                && checkPartOfSpeech(tbr.getVerb(), "verb")
+                && (tbr.getAdverb().equals(Constants.NULL_VALUE_WORD) || checkPartOfSpeech(tbr.getAdverb(), "adverb"))
+                && (tbr.getPreposition().equals(Constants.NULL_VALUE_WORD) || checkPartOfSpeech(tbr.getPreposition(), "preposition"));
 
         if (validPhrase) {
             tbr.setHasBeenGroomed(true);

```

Basically, removing the validatePhraseComponent() method, and adding the logic it had which logged the error message into checkPartOfSpeech().

mrsbluerose replied:
Refactored and removed validate Phrase component.

haxwell commented on src/main/java/com/savvato/tribeapp/services/ToBeReviewedCheckerServiceImpl.java:
```diff
@@ -5,110 +5,127 @@
         rejectedPhraseRepository.save(new RejectedPhrase(tbr.toString()));
-        reviewSubmittingUserRepository.deleteByToBeReviewedId(tbr.getId());
+        ReviewSubmittingUser rsu = new ReviewSubmittingUser(reviewSubmittingUserRepository.findUserIdByToBeReviewedId(tbr.getId()),tbr.getId());
+        reviewSubmittingUserRepository.delete(rsu);
```
Would deleteByToBeReviewedId() work here? That way, you wouldn't need line 127 and that call to get the user ID.. 

A delete by toBeReviewedId would delete this entry from the table for all users who submitted this phrase. I believe that is what we want, because we have just determined that this phrase is invalid. 

We do want some way of notifying each of those users that their phrase has been rejected. If you would, please add a TODO comment to the code to that affect, and please create an issue in Jira. Perhaps do those the other way around, and reference the Jira number in the TODO.

mrsbluerose replied:
The original code used a custom deletByToBeReviewedId. I kept getting the following errors:
- could not extract ResultSet; nested exception is org.hibernate.exception.GenericJDBCException: could not extract ResultSet
- Can not issue data manipulation statements with executeQuery()

I fixed it by calling using a ReviewSubmittingUser object and the hibernate .delete() method. I had to troubleshoot this for quite a while and blamed it on the fact that this database table uses a composite primary key. I am not sure why it didn't show up in earlier development, but I did have a cascading process of finding an error, fixing it and uncovering another. I would be happy to troubleshoot this in our meeting, if you want me to make the custom delete query work.

I will add the todo comment and ticket

haxwell replied:
I was wrong. I should have looked at the table. This is the correct way to do it. Never mind this comment.

haxwell reviewed (changes requested):
Great job so far! Your refactor shows an understanding of the code, the problem we're trying to solve and its logic. Your able to see the idea, and the places where we are describing it in code inefficiently. Very cool.

Some minor requested changes.

mrsbluerose commented:
Made additional changes:

1. refactored checkPartOfSpeech to include a try/catch for instance when we don't get a results set back from the API (we get an object, but it is not correctly configured with a "results" key in the json object)
2. added two tests for missing results set and missing part of speech

mrsbluerose commented on src/main/java/com/savvato/tribeapp/services/ToBeReviewedCheckerServiceImpl.java:
```diff
@@ -5,110 +5,129 @@
-        try {
-            if (validPartOfSpeech) {
+        } else {
+            JsonObject wordDetails = getWordDetails(word).get();
```
Saved to Optional.

mrsbluerose commented:
Removed unnecessary @Id annotation from the review submitting user repo as requested from our 8/29 meeting.
